### DISCUSSION TITLE
More lost than the average bear...

### DISCUSSION THREAD

**Discussion by pedz** [tailwindlabs/tailwindcss#9863] [Help]
I'm working on a Rails 7 project using Emacs.  I've done Rails projects in 2007 so I know a smattering of things but most of what I know is now obsolete.  So... what I "know" might actually be more of a hinderance.

I'm starting with adding the [Simple no labels Sign-in and Registration](https://tailwindui.com/components/application-ui/forms/sign-in-forms) form and I see tons of class names.  I have read Adam Wathan's "a few thousand words" so I get the gist of the class design.  I've also found a plug-in for Emacs to add Tailwind CSS IntelliSense but I haven't integrated it into my setup yet.

My question is how do I discover all the alternative class names?  For example, the web pages demonstrate tons of color choices and size choices, font choices, etc.  But how do I get a list of them?  It appears to me right now that my editor is suppose to be providing those choices.  Assuming I can **not** get something integrated into Emacs, what is the fall back method of learning about all the various class names and their intended use?

Another question: in the editor set up, it says that Prettier can automatically order the classes in the recommended way.  Prettier appears to be an npm "thing".  Rails 7 is attempting to be "node-less" and so currently npm isn't in my development environment and I know next to nothing about it.  So... again, is there a fall back method?  Is the recommended ordering of the classes documented somewhere?

Also... it may or may not make a difference in this particular questions but I did purchase the Refactoring UI course.  I've not dug into it yet except superficially but it seems to be higher level about UI general principles and not a "How to use Tailwind" tutorial -- which is fine.  But if all of the above is somewhere within the Refactoring UI course, then just point me there and I'll go exploring.

I plan to try and get the Emacs lsp-mode plug in to work which will provide me with Tailwind CSS IntelliSense and I'll update this question if I manage to get it working.

Thank you for your time

**Comment by wongjn** [comment]
> My question is how do I discover all the alternative class names? For example, the web pages demonstrate tons of color choices and size choices, font choices, etc. But how do I get a list of them? It appears to me right now that my editor is suppose to be providing those choices. Assuming I can not get something integrated into Emacs, what is the fall back method of learning about all the various class names and their intended use?

Use the [documentation website](https://tailwindcss.com/docs) to search for the property or class name candidate you'd like to use. After a while, you start to remember some or remember patterns without needing to consult documentation as much.

> Another question: in the editor set up, it says that Prettier can automatically order the classes in the recommended way. Prettier appears to be an npm "thing". Rails 7 is attempting to be "node-less" and so currently npm isn't in my development environment and I know next to nothing about it. So... again, is there a fall back method?

There is no fall back method to run prettier without npm that I know of.

> Is the recommended ordering of the classes documented somewhere?

There is a [Tailwind CSS best practices gist](https://gist.github.com/sandren/0f22e116f01611beab2b1195ab731b63) that covers some class ordering concepts.

**Reply by pedz** [reply]
> Use the [documentation website](https://tailwindcss.com/docs) to search for the property or class name candidate you'd like to use. After a while, you start to remember some or remember patterns without needing to consult documentation as much.

Ahh!  I arbitrarily looked up `mt-6` and I see it explains the logic of the naming system.  Yes... that helps a lot.
